src/test/java/io/vertx/test/core/tls/Trust.java

```diff
@@ -29,5 +43,90 @@
+    return new JksOptions().setValue(loadCAs(certMap, DUMMY_PASSWORD)).setPassword(DUMMY_PASSWORD);
+  };
+  Trust<TrustOptions> SNI_SERVER_ROOT_CA_AND_OTHER_CA_2 = () -> {
+    HashMap<String, String> certMap = new HashMap<>();
+    certMap.put("host3.com", "src/test/resources/tls/root-ca/ca-cert.pem");
+    certMap.put("host2.com", "src/test/resources/tls/other-ca/ca-cert.pem");
+    return new JksOptions().setValue(loadCAs(certMap, DUMMY_PASSWORD)).setPassword(DUMMY_PASSWORD);
+  };
+  Trust<TrustOptions> SNI_SERVER_ROOT_CA_FALLBACK = () -> {
+    HashMap<String, String> certMap = new HashMap<>();
+    certMap.put("xxx", "src/test/resources/tls/root-ca/ca-cert.pem");
+    return new JksOptions().setValue(loadCAs(certMap, DUMMY_PASSWORD)).setPassword(DUMMY_PASSWORD);
+  };
+  Trust<TrustOptions> SNI_SERVER_OTHER_CA_FALLBACK = () -> {
+    HashMap<String, String> certMap = new HashMap<>();
+    certMap.put("xxxx", "src/test/resources/tls/other-ca/ca-cert.pem");
+    return new JksOptions().setValue(loadCAs(certMap, DUMMY_PASSWORD)).setPassword(DUMMY_PASSWORD);
+  };
+
+  static Buffer loadCAs(Map<String, String> certMap, String password) {
+    KeyStore keyStore = null;
+    try {
+      keyStore = KeyStore.getInstance("jks");
+      keyStore.load(null, null);
+      int count = 0;
+      for (Map.Entry<String, String> entry : certMap.entrySet()){
+        Buffer certBuf = Buffer.buffer(Files.readAllBytes(Paths.get(entry.getValue()).toAbsolutePath()));
+        keyStore.setCertificateEntry(entry.getKey(), loadCerts(certBuf)[0]);
+      }
+      ByteArrayOutputStream out = new ByteArrayOutputStream();
+      keyStore.store(out, password.toCharArray());
+      return Buffer.buffer(out.toByteArray());
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  static List<byte[]> loadPems(Buffer data, String delimiter) throws IOException {
+    String pem = data.toString();
+    String beginDelimiter = "-----BEGIN " + delimiter + "-----";
+    String endDelimiter = "-----END " + delimiter + "-----";
+    List<byte[]> pems = new ArrayList<>();
+    int index = 0;
+    while (true) {
+      index = pem.indexOf(beginDelimiter, index);
+      if (index == -1) {
+        break;
+      }
+      index += beginDelimiter.length();
+      int end = pem.indexOf(endDelimiter, index);
+      if (end == -1) {
+        throw new RuntimeException("Missing " + endDelimiter + " delimiter");
+      }
+      String content = pem.substring(index, end);
+      content = content.replaceAll("\\s", "");
+      if (content.length() == 0) {
+        throw new RuntimeException("Empty pem file");
+      }
+      index = end + 1;
+      pems.add(Base64.getDecoder().decode(content));
+    }
+    if (pems.isEmpty()) {
+      throw new RuntimeException("Missing " + beginDelimiter + " delimiter");
+    }
+    return pems;
+  }
+
+  static X509Certificate[] loadCerts(Buffer buffer) throws Exception {
```

Comment:
isn't this code already in KeyStoreHelper ?

Reply:
Yes, these methods are more or less unchanged copies from __private__ methods in the KeyStoreHelper, which are not available for the tests. May be it would be nicer to share these methods with an Utility class. I just needed a way to create jks stores for the tests on the fly, also because it's self explaining how the key store is structured (the alias -> cert mapping).  Knowing the structure is important for understanding what the sni truststore test tests. Alternatively the keytool can be used to create the jks stores for the tests and add the files to the test resources. But I think then it's harder to get an idea what some of test do.

Reply:
currently keystores/truststores are created manually (check ssl.txt) and it is very fine this way, is that something we can use instead ?

Reply:
Yes sure. I will change it.

Reply:
done